### ISSUE TITLE
API proposal: `Property<T>`

### ISSUE THREAD

**Issue by nate-thegrate** [flutter/flutter#154197]
**TL;DR** at the bottom 🙂

<br>

# Problem

The current `WidgetStateProperty` API is fantastic—there's a [Decoding Flutter video](https://youtu.be/CylXr3AF3uU) with a great explanation of why—but it's also messy.

<br>

For every parameter that could potentially be a WidgetStateProperty, there are 3 possible types, as noted in [a recent Design Doc](https://flutter.dev/go/using-widget-state-property). Using the `Color` class as an example:

1. Some APIs only accept `Color`s
2. Some APIs only accept `WidgetStateProperty<Color>` objects
3. Some APIs can handle either a `Color` or a `WidgetStateProperty<Color>` being passed

<br>

Number 3 is possible thanks to an additional class declaration:

```dart
class WidgetStateColor extends Color implements WidgetStateProperty<Color> {
  // ...
}
```

Unfortunately, there's a long list of problems with classes like this one.

<br>

#### copy/paste typos

- https://github.com/flutter/flutter/pull/151935

We needed the PR shown above because repeatedly copy/pasting near-identical logic & documentation is a slog: it's boring to write and boring to review, so mistakes are more likely to slip through.

> [!NOTE]
> At the time of writing, https://github.com/flutter/flutter/pull/151935 has landed in master but not stable, so the [api.flutter.dev docs](https://api.flutter.dev/flutter/widgets/WidgetStateTextStyle-class.html) are still showing inaccurate documentation.

#### `const` constructors

```dart
/// To define a `const` [WidgetStateTextStyle], you'll need to extend
/// [WidgetStateTextStyle] and override its [resolve] method.
```

Technically, that was never true, since global functions and static methods can be referenced in a constant context:

```dart
class MyWidget extends StatelessWidget {
  const MyWidget({super.key});

  static TextStyle _favoriteStyle(Set<WidgetState> states) {
    if (states.contains(WidgetState.selected)) {
      return const TextStyle(color: Color(0xFF00FFFF));
    }
    return TextStyle(
      color: states.contains(WidgetState.disabled) ? Colors.black26 : Colors.black,
    );
  }

  @override
  Widget build(BuildContext context) {
    // defining a const WidgetStateTextStyle without extending the class!
    const widgetStateTextStyle = WidgetStateTextStyle.resolveWith(_favoriteStyle);

    // ...
  }
}
```

And on top of that, recently I was able to land a PR that adds another `const factory` constructor to WidgetStateTextStyle:

```dart
const widgetStateTextStyle = WidgetStateTextStyle.fromMap({
  WidgetState.focused:  TextStyle(color: Colors.blue, fontWeight: FontWeight.bold),
  WidgetState.disabled: TextStyle(color: Colors.grey),
  WidgetState.any:      TextStyle(color: Colors.black),
});
```

But even after the PR had been in review for over 2 months, neither I nor the reviewers noticed that the docs needed to be changed, since "why would anyone want to read through all of that?"

<br>

One more question about `const` constructors: why does `WidgetStateTextStyle` have a `const factory`, but `WidgetStateColor` doesn't?

```dart
  /// If used as a regular color, the color resolved in the default state (the
  /// empty set of states) will be used.
  factory WidgetStateColor.resolveWith(WidgetPropertyResolver<Color> callback) = _WidgetStateColor;
```

```dart
  /// If used as a regular text style, the style resolved in the default state (the
  /// empty set of states) will be used.
  const factory WidgetStateTextStyle.resolveWith(WidgetPropertyResolver<TextStyle> callback) = _WidgetStateTextStyle;
```

Answer: the [`WidgetStateTextStyle.resolveWith()`](https://main-api.flutter.dev/flutter/widgets/WidgetStateTextStyle/WidgetStateTextStyle.resolveWith.html) documentation is wrong.

```dart
final Color color = WidgetStateColor.resolveWith((_) => Colors.red);
print(color.value == Colors.red.value); // true

const emptyStyle = TextStyle();
const redText = TextStyle(color: Colors.red);
final TextStyle style = WidgetStateTextStyle.resolveWith((_) => redText);
print(style.color == redText.color); // false
print(style.color == emptyStyle.color); // true
```

<br>

## Unsafe types

The `num` class creates a beautiful type hierarchy.

<p align="center">
  <img 
    src="https://github.com/user-attachments/assets/dafef522-9bd9-4034-a2c4-3a2d92795ef0"
    alt="sealed num"
    width="50%"
  />
</p>

```dart
sealed class num {}
class int implements num {}
class double implements num {}
```

<br>

Imagine if this hierarchy were flipped upside down:

<p align="center">
  <img 
    src="https://github.com/user-attachments/assets/b5202cc6-517b-43f9-8cbc-3f96f7ecfe76"
    alt="upside down hierarchy"
    width="50%"
  />
</p>

```dart
class int {}
class double {}

/// The [num] class allows a [double] be used anywhere that accepts an [int],
/// but this should only be done with classes that explicitly support
/// both integer and non-integer values.
class num extends int implements double {}

class MyClass {
  const MyClass(this.decimal, this.integer, this.either);

  /// A floating-point decimal.
  final double decimal;

  /// You could pass a [num] here, but it might cause an error!
  final int integer;

  /// This can be either an integer or a decimal.
  /// Passing an [int] works fine, but for a [double],
  /// you have to type out `num(3.14)` in order to avoid a syntax error.
  final int either;
}
```

<br>

Replace `int`, `double`, and `num` with `Color`, `WidgetStateProperty<Color>` and `WidgetStateColor` respectively, and that's what we're currently working with.

From the `WidgetStateColor` docs:

```dart
/// [WidgetStateColor] should only be used with widgets that document
/// their support, like [TimePickerThemeData.dayPeriodColor].
```

<br>

It'd be great if, instead of `WidgetStateColor` / `WidgetStateTextStyle` / `WidgetStateMouseCursor` / `WidgetStateBorderSide`, we had 1 generic type to cover all the bases.

It'd also be great to do something about the type hierarchy:

<p align="center">
  <img 
    src="https://github.com/user-attachments/assets/b91038f2-82c0-4341-8b3c-2e0973ef4781"
    alt="WidgetStateColor hierarchy"
    width="75%"
  />
</p>

<br>

# Solution

```dart
typedef Property<T> = T | WidgetStateProperty<T>;
```

Much like how `num` is a 100% straightforward way to accept either an `int` or a `double`, the `Property` type would allow complete flexibility between e.g. `Color` and `WidgetStateProperty<Color>`.

Currently, we have 2 ways to create a `ButtonStyle`:

```dart
const ButtonStyle({
  WidgetStateProperty<Color?>? backgroundColor,
  WidgetStateProperty<Color?>? foregroundColor,
});

static ButtonStyle styleFrom({
  Color? backgroundColor,
  Color? foregroundColor,
}) {
  // ...
}
```

But `Property<T>` makes it easy to consolidate:

```dart
const ButtonStyle({
  Property<Color?>? backgroundColor,
  Property<Color?>? foregroundColor,
})
```

This has 2 huge benefits:
1. Along with the several flaws listed above, `WidgetStateColor` implements the `WidgetStateProperty<Color>` interface, so it doesn't have the added flexibility of a nullable `WidgetStateProperty<Color?>`. Instead of defining a new `WidgetStateNullableColor` type, we can just add a `?` to the existing type argument.
2. This creates a meaningful distinction between `T` and `WidgetStatePropertyAll<T>`:
  a. `Colors.blue`: make it blue, but default to gray if disabled
  b. `const WidgetStatePropertyAll(Colors.blue)`: make it blue no matter what

<br>

### Caveat

This solution doesn't actually work.

```dart
abstract interface class Property<T> {}

class WidgetStateProperty<T> implements Property<T> {}

class num implements Property<num> {}
class Color implements Property<Color> {}
class ShapeDecoration implements Property<ShapeDecoration> {}
```

The ability to specify a union type `Property<T> = T | WidgetStateProperty<T>` is amazing, since it implicitly applies to every class declaration. However, as of yet this is not supported by Dart syntax.

- https://github.com/dart-lang/language/issues/83

If somebody is interested in working with the Dart team to implement this language feature, that'd be awesome. Or if someone on the Dart team opens up a Patreon and giving small quantities of money each month helps to expedite the "union types" feature, then I'd be down!

<br><br>

**TL;DR**: the goal here is to deprecate & remove `WidgetStateColor`, `WidgetStateBorderSide`, `WidgetStateTextStyle`, etc. in favor of a more useful generic type. The ideal solution would be something like `typedef Property<T> = T | WidgetStateProperty<T>`
…but it requires [a language feature](https://github.com/dart-lang/language/issues/83) that we don't currently have.

**Comment by MitchellGoodwin** [comment]
> Or if someone on the Dart team opens up a Patreon and giving small quantities of money each month helps to expedite the "union types" feature, then I'd be down!

I understand the energy, but if we start exchanging money to get features done in Flutter, that would be a whole can of worms. On top of legal issues, we don't want to create a literal bidding war to get certain features in.

**Comment by nate-thegrate** [comment]
Perhaps one could argue that a potential for legal issues would be more of a hurdle to get past than something that should shut the idea down.

But overall I agree that it's probably best to avoid rocking the boat 😁

**Comment by mateusfccp** [comment]
Why can't it be implemented with a tagged union type?

I know the ergonomics of wrapping the values is not the best, but wouldn't the general problem be solved?

**Comment by nate-thegrate** [comment]
```dart
// current
return Material(color: _backgroundColor, shadowColor: Colors.black45, elevation: 5.0);

// with tagged union
return Material(
  color: Property(_backgroundColor),
  shadowColor: const Property(Colors.black45),
  elevation: const Property(5.0),
);
```

I imagine that switching to tagged union types would be a great way to reduce the number of people who want to use Flutter.

**Comment by mateusfccp** [comment]
I don't know about this, I just proposed a solution that fixes the problem without requiring a change to the language.

By using this approach, the state goal ("to deprecate & remove ... in favor of a more useful generic type") is achievable.

This makes false the statement that "it requires a language that we don't currently have".

If this approach is worse or better than what we have now, is another discussion. I, personally, wouldn't have problem using it, but I agree that the ergonomics could be improved if we had other feature in the language level.

**Comment by nate-thegrate** [comment]
I believe we share the same understanding that this is indeed an ergonomic problem, apologies for not being explicit in my agreement about that earlier.

But when someone runs `flutter upgrade` only to realize that half of the entire codebase now needs every argument wrapped in another object constructor, the word "problem" starts to seem like too gentle a term to describe the situation.

**Comment by MitchellGoodwin** [comment]
Personally, I don't know if I like the idea of using the `Property<T>` style for union types. I'm assuming the idea is not to make `Property` a keyword and anything could be used there? We already use the `foo<T>` definition for some things, like routes, and I think it'd be better if there was a more clear signal at a glance that the type requested is a union type.

**Comment by nate-thegrate** [comment]
> Personally, I don't know if I like the idea of using the `Property<T>` style for union types.

That's fair.

As an alternative to `Property<Color>`, we could follow the precedent of `FutureOr<T>` and make it `WidgetStatePropertyOr<Color>`, though that's quite a bit more verbose.

Or, since we already have `WidgetPropertyResolver<T>` without the "state" in the name, the union type name could be `WidgetPropertyOr<T>`.

**Comment by eernstg** [comment]
@nate-thegrate, that's a great analysis!

A couple of comments on the unions:

Tagged unions are generally costly at run time because we have to have an object that has the union type and determines which operand of the union is the type of the actual value. Also, as mentioned in several comments above, we can't obtain the desired subtyping relations (where `T1 <: T1 | T2` and `T2 <: T1 | T2`, `T1 | T2 == T2 | T1`, and so on).

Other than that, a tagged union type solution could be expressed as follows:

```dart
// Minimal tagged union type emulation.

class Union2<T1, T2> {
  final Object? _value;
  Union2.in1(T1 this._value);
  Union2.in2(T2 this._value);

  bool get is1 => _value is T1;
  bool get is2 => _value is T2;
  T1 get as1 => _value as T1;
  T2 get as2 => _value as T2;
}

// Example.

class Color {
  final String name;
  Color(this.name);
}

class Property<T> {
  // `resolve` would produce a `T`, but for this short example we'll just
  // show the type by returning a `String`.
  String resolve() => T.toString();
}

void acceptColor(Color color) {}

void acceptProperty<T>(Property<T> property) {}

void acceptBoth(Union2<Color, Property<Color>> both) {
  if (both.is1) {
    var value = both.as1;
    print('A ${value.name} color');
  } else if (both.is2) {
    var value = both.as2;
    print('A property that resolves as a ${value.resolve()}');
  }
}

void main() {
  var color = Color('green');
  var property = Property<Color>();
  var both1 = Union2<Color, Property<Color>>.in1(color);
  var both2 = Union2<Color, Property<Color>>.in2(property);

  // Matching cases.
  acceptColor(color);
  acceptProperty(property);
  acceptBoth(both1);
  acceptBoth(both2);

  // Cases where a match can be expressed.
  acceptBoth(Union2.in1(color));
  acceptBoth(Union2.in2(property));

  // Cases that can't and shouldn't match.
  //   acceptColor(property); // Error.
  //   acceptProperty(color); // Error.
}
```

Note that we have to invoke a constructor (`Union2.in1` or `Union2.in2`) in order to obtain an expression whose type is the union type, because we don't have subtype relations like `S <: Union2<S, T>` or `T <: Union2<S, T>`.

As mentioned above, implicit constructors could perform this transformation implicitly.

However, if you're happy about a representation that has a couple of extra constraints then you can avoid the run-time overhead by using extension types to model the union:

```dart
import 'package:extension_type_unions/extension_type_unions.dart';

class Color {
  final String name;
  Color(this.name);
}

class Property<T> {
  // `resolve` produces a `T`, but for this short example we'll just
  // show the type by returning a `String`.
  String resolve() => T.toString();
}

void acceptColor(Color color) {}
void acceptProperty<T>(Property<T> property) {}

void acceptBoth(Union2<Color, Property<Color>> both) => both.split(
  (color) => print('A ${color.name} color'),
  (property) => print('A property that resolves as a ${property.resolve()}'),
);

void main() {
  var color = Color('green');
  var property = Property<Color>();

  // For expressions whose type is the union, we can call a constructor
  // (which is verbose) or we can use `uNM` where `N` is the number of
  // operands of the union and `M` is the selected operand for this value.
  // So `e.u22` is a `Union2<T1, T2>` where `T2` is the static type of `e`.
  var both1 = Union2<Color, Property<Color>>.in1(color);
  var both2 = property.u22;

  // Matching cases.
  acceptColor(color);
  acceptProperty(property);
  acceptBoth(both1);
  acceptBoth(both2);

  // Cases where a match can be expressed.
  acceptBoth(color.u21);
  acceptBoth(property.u22);

  // Cases that can't and shouldn't match.
  //   acceptColor(property); // Error.
  //   acceptProperty(color); // Error.
}
```

The extension type "wrapper" has a cost of zero at run time (because it doesn't exist, and the constructor invocation can be compiled away entirely). On the other hand, there's no run-time enforcement of the typing, we only get the compile-time type checks. (It is even possible to do `true as Union2<int, String>`.)

If you're happy about relying on compile-time checks only then you can express union types in Dart of today in this manner.

The final part which is missing (I'd assume: from every proposal other than real union types) is that none of these proposals have the desired subtype relationships, which again means that you have to do something in order to obtain an expression whose type is a union type.

In this case it is sufficient to add `.uNM` to the given expression, which will have the same value, but the type will be a union type with `N` operands, and the given value is going to have the type of operand `M` (for example, `1.u21` has type `Union2<int, Never>` and `'Hello!'.u33` has type `Union3<Never, Never, String>`).

**Comment by nate-thegrate** [comment]
@eernstg thanks for the informative response.

<br>

My personal thoughts, which others on the Flutter team may or may not agree with:

> The extension type "wrapper" has a cost of zero at run time (because it doesn't exist, and the constructor invocation can be compiled away entirely). On the other hand, there's no run-time enforcement of the typing, we only get the compile-time type checks. (It is even possible to do `true as Union2<int, String>`.)

That sounds fantastic. I bet the consensus among app developers would be to go for the option with the best performance, even if it allows for edge cases where debugging might be more difficult. (One solution is: if you don't want errors caused by `dynamic` or `foo as SomeClass`, then don't use `dynamic` or `foo as SomeClass`!)

<br><br>

I suppose even without implicit constructors or the extension_type_unions package, we could implement something like this today:

```dart
class MyButton extends StatelessWidget {
  const MyButton({this.color});
  final Property<Color?>? color;

  // ...
}

extension FromAny<T> on T {
  Property<T> get p => ValueProperty(this);
}
```

```dart
class SomeWidget extends StatelessWidget {
  final Color? color;
  final WidgetStateProperty<Color?>? wspColor;

  // ...

  Widget build(BuildContext context) {
    return Column(children: [
      MyButton(color: wspColor),
      MyButton(color: color.p),
    ]);
  }
}
```

Having `color.p` in place of `Property(color)` would be much nicer, though I am worried about a large potential drawback:

```dart
const MyButton(color: Colors.red.p); // error
```

My understanding is that regardless of whether extension types are used, choosing not to wrap a value in a `Property()` constructor could lead to a significant performance setback that could only be prevented by an `implicit const` constructor or something similar.

**Comment by nate-thegrate** [comment]
> Wouldn't `.p` be very obscure? It would also disregard the style guide about [avoiding abbreviations](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo/985001f5b97164e3d87ab4f97188622de43e3d23#avoid-abbreviations).
> 
> The alternative would be to clearly state `a.property`, but at this point It's just 1 character less than `property(a)`, so I'm not sure if it's worthy.

True that!

**Comment by eernstg** [comment]
I don't know if this is already obvious, but let me clarify a few points.

I'm assuming that some declarations can be modified (in particular, I believe that some `WidgetStateProperty` related declarations should be modified). If that's already too much of a breaking change then this approach won't fly.

@nate-thegrate wrote:
> The current `WidgetStateProperty` API is fantastic—there's a [Decoding Flutter video](https://youtu.be/CylXr3AF3uU) with a great explanation of why—but it's also messy.
> 
> 
> For every parameter that could potentially be a WidgetStateProperty, there are 3 possible types, as noted in [a recent Design Doc](https://flutter.dev/go/using-widget-state-property). Using the `Color` class as an example:
> 
> 1. Some APIs only accept `Color`s
> 2. Some APIs only accept `WidgetStateProperty<Color>` objects
> 3. Some APIs can handle either a `Color` or a `WidgetStateProperty<Color>` being passed

Using extension type unions (or any kind of union types, for that matter), we'd handle that by having a parameter type of (1) `Color`, (2) `WidgetStateProperty<Color>`, and (3) `WidgetStatePropertyOr<Color>`, which is declared as follows:

```dart
import 'package:extension_type_unions/extension_type_unions.dart';

typedef WidgetStatePropertyOr<X> = Union2<WidgetStateProperty<X>, X>;
```

The declaration of `WidgetStateColor` might not be needed at all, because there's no reason (that I know of) to have a type which is both sides of the union, we just use one or the other, and some formal parameters can accept both.

It's actually an improvement if we don't have to remember to check whether any given `Color` object is actually a `WidgetStateColor` whose correct treatment would include the invocation of `resolve` at specific occasions.

> Number 3 is possible thanks to an additional class declaration:
> 
> ```dart
> class WidgetStateColor extends Color implements WidgetStateProperty<Color> {
>   // ...
> }
> ```

This declaration might go away entirely, or we might remove `extends Color`: A `WidgetStateProperty<Color>` can be passed to an argument of type `WidgetStateProperty<Color>`, and it can be passed to an argument of type `WidgetStatePropertyOr<Color>` via the "injection" getter `u21`. It _should not_ be possible to pass it to a parameter of type `Color`.

I don't know all the details, but it seems likely to me that most of the complexity of all the subtypes of `WidgetStateProperty<T>` is associated with being able to `resolve` the property to a `T` and also being able to work as a `T`.

This is not needed if `WidgetStateTextStyle` is not a subtype of `TextStyle` (and similarly for other subtypes of `WidgetStateProperty`), and it doesn't have to be a subtype of `TextStyle` when the formal parameter types will accept a `WidgetStateProperty<...>` or a `WidgetStatePropertyOr<...>`.

> #### `const` constructors
> ```dart
> /// To define a `const` [WidgetStateTextStyle], you'll need to extend
> /// [WidgetStateTextStyle] and override its [resolve] method.
> ```
>
> Technically, that was never true, ...

In any case, it would be _easier_ to do anything, including writing a constant constructor, when the relevant classes are made simpler (because they will only play one of the two roles in the union, not both of them).

> ## Unsafe types
> ...
> 
> From the `WidgetStateColor` docs:
> ...
> ```dart
> /// [WidgetStateColor] should only be used with widgets that document
> /// their support, like [TimePickerThemeData.dayPeriodColor].
> ```
> 
> 
> It'd be great if, instead of `WidgetStateColor` / `WidgetStateTextStyle` / `WidgetStateMouseCursor` / `WidgetStateBorderSide`, we had 1 generic type to cover all the bases.
> 
> It'd also be great to do something about the type hierarchy:
> 
> <img alt="WidgetStateColor hierarchy" width="75%" src="https://private-user-images.githubusercontent.com/10457200/361935342-4d864e2a-3e78-4061-b6ca-429a11c0780f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ5NjEzMzgsIm5iZiI6MTcyNDk2MTAzOCwicGF0aCI6Ii8xMDQ1NzIwMC8zNjE5MzUzNDItNGQ4NjRlMmEtM2U3OC00MDYxLWI2Y2EtNDI5YTExYzA3ODBmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI5VDE5NTAzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjYzg2M2FmOTQyNGEwMTZlM2UyZmRmMjk2ZTYxOTMzNWJiMGJlZDExMmVjNDI3Yzc1NjNiZmIzY2Y5NGRjY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.FS5kolFRRTtUsNLX8jcuERJls0LZRl2UwpeHQ8kCftw">
>

With extension type unions we'd have fewer subtype relationships:

```mermaid
graph BT;
  c[Color]
  wsc[WidgetStateColor] --> wspc["WidgetStateProperty<Color>"]
  both[WidgetStatePropertyOr<Color>]
```

We do not have all the desired subtype relationships (and `WidgetStateColor` may have been eliminated entirely), but we are using an injection (`u21` to map a property to the union, and `u22` to map the underlying type like `Color` to the union).

> # Solution
> ```dart
> typedef Property<T> = T | WidgetStateProperty<T>;
> ```

This is what we're modeling with `WidgetStatePropertyOr<T>`.

> Much like how `num` is a 100% straightforward way to accept either an `int` or a `double`, the `Property` type would allow complete flexibility between e.g. `Color` and `WidgetStateProperty<Color>`.

With extension type unions, we'd use `myProperty.u21` and `myColor.u22` to obtain a `WidgetStatePropertyOr<Color>`. That isn't 100% straightforward, but with implicit constructors we could write `myProperty` or `myColor` in some situation where the context type is the union, and they would then implicitly be transformed into `WidgetStatePropertyOr<Color>.in1(myProperty)` respectively `WidgetStatePropertyOr<Color>.in2(myColor)` (which is essentially the same thing as `myProperty.u21` and `myColor.u22`).

> Currently, we have 2 ways to create a `ButtonStyle`:
> 
> ```dart
> const ButtonStyle({
>   WidgetStateProperty<Color?>? backgroundColor,
>   WidgetStateProperty<Color?>? foregroundColor,
> });
> 
> static ButtonStyle styleFrom({
>   Color? backgroundColor,
>   Color? foregroundColor,
> }) {
>   // ...
> }
> ```
> 
> But `Property<T>` makes it easy to consolidate:
> 
> ```dart
> const ButtonStyle({
>   Property<Color?>? backgroundColor,
>   Property<Color?>? foregroundColor,
> })
> ```

This would become

```dart
const ButtonStyle({
  WidgetStatePropertyOr<Color?> backgroundColor,
  WidgetStatePropertyOr<Color?> foregroundColor,
});
```

One thing is missing here: If the value of `backgroundColor` is null then it is interpreted as either a `Color?` whose value is null, or as a `WidgetStateProperty<Color?>` whose color is null. There is no way we can say for sure which one it is.

If it is necessary to maintain this distinction (that is, we can't just say that it always means "a color which isn't there", as a convention) then we'll have to use a real wrapper object (that is, we would not use an extension type, we'd use a regular class to model the union types, and pay for the extra time/space in order to be able to make this distinction).

> This has 2 huge benefits:
> 
> 1. Along with the several flaws listed above, `WidgetStateColor` implements the `WidgetStateProperty<Color>` interface, so it doesn't have the added flexibility of a nullable `WidgetStateProperty<Color?>`. Instead of defining a new `WidgetStateNullableColor` type, we can just add a `?` to the existing type argument.

We can use a convention to express this using extension type unions, but we can't make the property nullable as a whole (or, at least, we can't see which case we have if we encounter the value null).

> 2. This creates a meaningful distinction between `T` and `WidgetStatePropertyAll<T>`:
>    a. `Colors.blue`: make it blue, but default to gray if disabled
>    b. `const WidgetStatePropertyAll(Colors.blue)`: make it blue no matter what

This shouldn't be hard to express: The former would be an actual `Color`, the latter would be an instance of a subclass of `WidgetStateProperty<Color>` that stores a `Color` and uses it as would be expected for the `...All` variant.
 

> Or, another great option would be an "implicit" constructor:

Yes, that would be convenient, especially for unions.

> **TL;DR**: the goal here is to deprecate & remove `WidgetStateColor`, `WidgetStateBorderSide`, `WidgetStateTextStyle`, etc. in favor of a more useful generic type. ...

I'd hope that, too.

**Comment by nate-thegrate** [comment]
Yeah, overall this looks really good!

> One thing is missing here: If the value of `backgroundColor` is null then it is interpreted as either a `Color?` whose value is null, or as a `WidgetStateProperty<Color?>` whose color is null. There is no way we can say for sure which one it is.

I'm glad this was brought up, and fortunately I don't think it'll be an issue. In almost any case, `null` would mean "I don't want to change the default", so whether it's conceptualized as "null Color" or "null WidgetStateProperty" isn't super important.

<br>

> with implicit constructors we could write `myProperty` or `myColor` in some situation where the context type is the union, and they would then implicitly be transformed into `WidgetStatePropertyOr<Color>.in1(myProperty)` respectively `WidgetStatePropertyOr<Color>.in2(myColor)` (which is essentially the same thing as `myProperty.u21` and `myColor.u22`).

That sounds great. If we had:

```dart
typedef WidgetStatePropertyOr<X> = Union2<WidgetStateProperty<X>, X>;
```

which worked the same as `sealed class WidgetStatePropertyOr<X>` with better performance, I'd be all for it.

<br><br>

But I do think implicit constructors are a requirement for Flutter's use case…

```dart
// before
return const ExpensiveStatelessWidget(color: Colors.red);

// after
return ExpensiveStatelessWidget(color: Colors.red.u22);
```

…since the "after" example here could be seen as a significant downgrade.

**Comment by eernstg** [comment]
> But I do think implicit constructors are a requirement for Flutter's use case…

Vote for https://github.com/dart-lang/language/issues/108, and tell others to do so, if they care about this. It does matter whether any particular proposal has 5 votes or 5000.
